From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Rong Bao <rong.bao@csmantle.top>
Date: Thu, 19 Feb 2026 21:32:17 +0800
Subject: [PATCH 50/50] CSMANTLE: LoongArch: Add CONFIG_LOONGARCH_NATIVE_CPU

All of GCC, Clang, and rustc supports native architecture detection on
LoongArch platforms. Enabling this may improve instruction selection and
static scheduling performance.

The Kconfig and Makefile organization is learned from x86's config
files.

Signed-off-by: Rong Bao <rong.bao@csmantle.top>
---
 arch/loongarch/Kconfig  | 24 ++++++++++++++++++++++++
 arch/loongarch/Makefile |  5 +++++
 2 files changed, 29 insertions(+)

diff --git a/arch/loongarch/Kconfig b/arch/loongarch/Kconfig
index a9b26a8e6f0a36b26aadab367da042a1e94cceca..ab8d8e7406c5606f7f5f68b946ac5d20a9378fd8 100644
--- a/arch/loongarch/Kconfig
+++ b/arch/loongarch/Kconfig
@@ -339,6 +339,17 @@ config AS_HAS_LVZ_EXTENSION
 config AS_HAS_SCQ_EXTENSION
 	def_bool $(as-instr,sc.q \$t0$(comma)\$t1$(comma)\$t2)
 
+# From arch/x86/Kconfig.cpu
+config CC_HAS_MARCH_NATIVE
+	# This flag might not be available in cross-compilers:
+	def_bool $(cc-option, -march=native)
+	# LLVM 18 has an easily triggered internal compiler error in core
+	# networking code with '-march=native' on certain systems:
+	# https://github.com/llvm/llvm-project/issues/72026
+	# LLVM 19 introduces an optimization that resolves some high stack
+	# usage warnings that only appear wth '-march=native'.
+	depends on CC_IS_GCC || CLANG_VERSION >= 190100
+
 config CC_HAS_ANNOTATE_TABLEJUMP
 	def_bool $(cc-option,-mannotate-tablejump)
 
@@ -678,6 +689,19 @@ config CPU_HAS_PREFETCH
 	bool
 	default y
 
+config LOONGARCH_NATIVE_CPU
+	bool "Build and optimize for local/native CPU"
+	depends on CC_HAS_MARCH_NATIVE
+	help
+	  Optimize for the current CPU used to compile the kernel.
+	  Use this option if you intend to build the kernel for your
+	  local machine.
+
+	  Note that such a kernel might not work optimally on a
+	  different LoongArch machine.
+
+	  If unsure, say N.
+
 config ARCH_SUPPORTS_KEXEC
 	def_bool y
 
diff --git a/arch/loongarch/Makefile b/arch/loongarch/Makefile
index 4b22f95aaafb62cd6c2504dc73bc65c3bbdf32a8..70a7dcf662d0d1d29771b96057cf5d281a2d9473 100644
--- a/arch/loongarch/Makefile
+++ b/arch/loongarch/Makefile
@@ -140,6 +140,11 @@ KBUILD_LDFLAGS			+= $(call ld-option,-mllvm --loongarch-annotate-tablejump)
 endif
 endif
 
+ifdef CONFIG_LOONGARCH_NATIVE_CPU
+KBUILD_CFLAGS += -march=native
+KBUILD_RUSTFLAGS += -Ctarget-cpu=native
+endif
+
 KBUILD_RUSTFLAGS		+= --target=loongarch64-unknown-none-softfloat -Ccode-model=small
 KBUILD_RUSTFLAGS_KERNEL		+= -Zdirect-access-external-data=yes
 KBUILD_RUSTFLAGS_MODULE		+= -Zdirect-access-external-data=no
-- 
2.55.0

